client/client-lib/src/modules.rs

```diff
@@ -0,0 +1,260 @@
+    inputs: Vec<Input>,
+    outputs: Vec<Output>,
+    signature: Signature,
+}
+
+impl UnsignedTransaction {
+    /// Generate change outputs using `module`
+    fn add_change(
+        &mut self,
+        module: &FedimintModule,
+        _rng: &mut DeterministicRandomnessTracker,
+    ) -> Result<Vec<SpendableOutput>, TransactionError> {
+        todo!()
+    }
+
+    /// Sign into [`Transaction`] and return new [`SpendableOutput`]s
+    fn sign(
+        &self,
+        _rng: &mut DeterministicRandomnessTracker,
+    ) -> (Transaction, Vec<SpendableOutput>) {
+        // check if inputs == outputs
+        // validate input sigs?
+        // calculate txid
+        // convert all the `SpendableOutput`s to `Inputs`
+        // use keys to sign tx
+        todo!()
+    }
+}
+
+pub trait IFedimintModule {
+    fn module_key(&self) -> &'static str;
+
+    /// Generate a transaction [`Output`] and
+    fn generate_output(
+        &self,
+        amount: Amount,
+        deterministic_rand: &bitcoin::hashes::sha256::Hash,
+    ) -> (Output, SpendableOutput);
+
+    fn decode_asset(&self, d: &mut dyn io::Read) -> Result<SpendableOutput, DecodeError>;
+}
+
+#[derive(Clone)]
+#[autoimpl(Deref using self.0)]
+pub struct FedimintModule(Arc<dyn IFedimintModule + Send + Sync + 'static>);
+
+// autoimpl below expands to something like:
+// impl Deref for MintModule {
+//     type Target = dyn IMintModule + Send + Sync + 'static;
+
+//     fn deref(&self) -> &Self::Target {
+//         &*self.0
+//     }
+// }
+
+pub struct MintModule {
+    config: MintModuleConfig,
+}
+
+pub struct MintModuleConfig {
+    some_number: u32,
+}
+
+impl MintModule {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn from_config(config: MintModuleConfig) -> FedimintModule {
+        FedimintModule(Arc::new(Self { config }))
+    }
+}
+
+impl IFedimintModule for MintModule {
+    fn module_key(&self) -> &'static str {
+        "mint"
+    }
+
+    fn generate_output(
+        &self,
+        _amount: Amount,
+        _deterministic_rand: &bitcoin::hashes::sha256::Hash,
+    ) -> (Output, SpendableOutput) {
+        todo!()
+    }
+
+    fn decode_asset(&self, _d: &mut dyn io::Read) -> Result<SpendableOutput, DecodeError> {
+        todo!()
+    }
+}
+
+/// Something that keeps tracks of which nonces (and any other deterministic randomness we need)
+///
+/// Details to be flushed out.
+pub struct DeterministicRandomnessTracker;
+
+#[derive(Default)]
+pub struct Client {
+    modules: BTreeMap<&'static str, FedimintModule>,
+}
+
+#[derive(Debug, Error)]
+pub enum DBError {
+    #[error("not found")]
+    NotFound,
+}
+
+impl Client {
+    pub fn register_module(&mut self, module: FedimintModule) -> &mut Self {
+        self.modules
+            .insert(module.module_key(), module)
+            .expect("Must not register modules with key conflict");
+        self
+    }
+
+    pub fn load_module_assets(&self, _module_key: &'static str) -> Result<Vec<Vec<u8>>, DBError> {
```

Comment:
We might need to prefix values/keys stored in a database with a "module key", or use different prefixes or something.

There are possibly other ways to approach it, but since `Client` can't do deserialization on it's own, it needs to be able to tell to which module each value belongs,. and then use the module to deserializa the object it wants.